### ISSUE TITLE
web config limits the number of characters usable to name secrets

### ISSUE THREAD

**Issue by sthaha** [prometheus-operator/prometheus-operator#4944]
**What happened?**

To configure TLS, certs and key is added to a `secret`. E.g. for `Prometheus` CR is named `project-appname-prometheus` its TLS assets are in a secret (injected using secrets field) - `projectname-appname-prometheus-tls-secret` 

When the secret is injected into prometheus using the follow spec 
```
spec: 
   secrets:
   - projectname-appname-prometheus-tls-secret
   web:
    tlsConfig:
      cert:
        secret:
          key: tls.crt
          name:  projectname-appname-prometheus-tls-secret
```

The stateful set generated by the operator has `volume.name` as `web-config-tls-secret-key-projectname-appname-prometheus-tls-secret` which is more than 63 chars allowed by k8s volume name ([see label names - RFC 1123 Label Names](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names) )

**Did you expect to see something different?**

yes,  the operator should generate a valid statefulset (by generating a label name that complies with volume name scheme ) or  refuse to accept the resource (validating webhook). 

**How to reproduce it (as minimally and precisely as possible)**:
1. Create a secret with name `projectname-appname-prometheus-tls-secret` that has an empty `ca.crt` and `key` 
2. configure prometheus to use web-config for tls
3. apply the prometheus resource 
4. See reasons for pod failing

 **Workaround** 

* Create secret name lesser than 40 chars by taking into account [the additional `web-config-tls-` that gets added to the secret name when PO generates volume.name](https://github.com/prometheus-operator/prometheus-operator/blob/363a4a7dfcd1e20c3a4737a6ba3f45e03a6d702b/pkg/webconfig/tls_credentials.go#L61)

**Environment**

Prometheus Operator version:

* Prometheus Operator version:   `v0.57.0` (but the same should happen with `v0.58.0`)

  
* Kubernetes version information: `v1.24`

* Kubernetes cluster kind: `OpenShift`

* Manifests:

N/A

* Prometheus Operator Logs:

 N/A 

**Anything else we need to know?**:

**Comment by simonpasquier** [comment]
IIUC this affects potentially all secrets/configmaps managed by the operator so we need a generic solution that shortens the name of the resource if needed.

**Comment by sthaha** [comment]
That's right @simonpasquier 
I was thinking of using a short sha1 for the name e.g. 
`web-config-tls-secret-key-projectname-appname-prometheus-tls-secret` would become `web-tls-secret-<sha1 (key + secret-name)[8]` for the volume names

**Comment by simonpasquier** [comment]
Sounds good to me. A small improvement could be to truncate `web-config-tls-secret-key-projectname-appname-prometheus-tls-secret` so that it fits the 63 chars limit with the SHA suffix. E.g. `web-config-tls-secret-key-projectname-appname-prometheus-tls-secret` would be turned into `web-config-tls-secret-key-projectname-appname-promethe-89c72073`.

**Comment by sthaha** [comment]
@simonpasquier could you please assign the bug to me?

**Comment by simonpasquier** [comment]
Closed by #4822 4988
